config/cmake/HDF5DeveloperBuild.cmake

```diff
@@ -0,0 +1,207 @@
+#
+# Copyright by The HDF Group.
+# All rights reserved.
+#
+# This file is part of HDF5.  The full HDF5 copyright notice, including
+# terms governing use, modification, and redistribution, is contained in
+# the COPYING file, which can be found at the root of the source code
+# distribution tree, or in https://www.hdfgroup.org/licenses.
+# If you do not have access to either file, you may request a copy from
+# help@hdfgroup.org.
+#
+
+# CMake settings for HDF5 Developer mode builds
+
+# Set CMake C++ flags based off of Debug build flags
+set (CMAKE_CXX_FLAGS_DEVELOPER ${CMAKE_CXX_FLAGS_DEBUG} CACHE STRING
+  "Flags used by the C++ compiler during developer builds." FORCE
+)
+
+# Set CMake C flags based off of Debug build flags. Add in -Og
+# option to disable some GCC optimizations that might affect
+# debugging negatively and also include some GCC compiler passes
+# that collect debugging information
+set (CMAKE_C_FLAGS_DEVELOPER "${CMAKE_C_FLAGS_DEBUG} -Og" CACHE STRING
+  "Flags used by the C compiler during developer builds." FORCE
+)
+
+# Set CMake binary linker flags based off of Debug binary linker flags
+set (CMAKE_EXE_LINKER_FLAGS_DEVELOPER ${CMAKE_EXE_LINKER_FLAGS_DEBUG}
+  CACHE STRING "Flags used for linking binaries during developer builds."
+  FORCE
+)
+
+# Set CMake shared library linker flags based off of Debug shared library
+# linker flags
+set (CMAKE_SHARED_LINKER_FLAGS_DEVELOPER ${CMAKE_SHARED_LINKER_FLAGS_DEBUG}
+  CACHE STRING "Flags used by the shared libraries linker during developer builds."
+  FORCE
+)
+
+mark_as_advanced (
+  CMAKE_CXX_FLAGS_DEVELOPER
+  CMAKE_C_FLAGS_DEVELOPER
+  CMAKE_EXE_LINKER_FLAGS_DEVELOPER
+  CMAKE_SHARED_LINKER_FLAGS_DEVELOPER
+)
+
+#-----------------------------------------------------------------------------
+# Define various HDF5 macros for debugging the library
+#-----------------------------------------------------------------------------
+
+# Enable debugging of various HDF5 modules
+set (HDF5_ENABLE_DEBUG_APIS ON CACHE BOOL "Turn on extra debug output in all packages" FORCE)
+
+# Enable debugging of API contexts
+add_compile_definitions (H5CX_DEBUG)
```

Comment:
These defines are only needed for the library, right? 
see the src/CMakeLists.txt file about line 1183; this "target_compile_definitions" is preferred. In this case, create a HDF5_ENABLE_DEBUG_DEVEL_APIS option and then in src/CMakeLists.txt add annother line to the target.

Reply:
If we don't consider other future debug options (like for the tools), then yes these are only relevant to the library, but there's a bit of an organization issue. I'd like to use a specific target_compile_definitions, but couldn't here because the target for the library isn't defined yet at the point where this .cmake file is included in the top-level CMakeLists.txt. I want to set all the debug options as early as possible so they apply to the whole project and are also defined before we start building the library and its components. If there's a way we can ensure the library's target is defined in the top-level CMakeLists.txt, then we could reduce the scope that these definitions are applied at.

As for the HDF5_ENABLE_DEBUG_DEVEL_APIS, there's already an option for this; the `set` call is just the best way I could see of force-overriding what was previously set for that option. I could have created another variable, but I didn't want to have to check 2 things before deciding whether to enable the debug APIs. I just wanted to re-use what already exists. It seems a bit unclean since I apparently have to specify the same docstring that the option was created with, but let me know if there's a better wayt to do this in CMake.

Reply:
set call is fine - it's the add_compile_definitions - by adding the BOOL generator lines to the target they are only applied to that target with the correct visibility. If you want the same defs applied to another target - you just add the same line to those targets. Only those targets you add the line to is affected.

Reply:
CMake is pushing target level adjustments rather then global adjustments.

Reply:
I probably need to start using generators more, since more CMake commands are supporting it.

Reply:
> set call is fine - it's the add_compile_definitions - by adding the BOOL generator lines to the target they are only applied to that target with the correct visibility. If you want the same defs applied to another target - you just add the same line to those targets. Only those targets you add the line to is affected.

Why create another yet another generator line for the HDF5 lib target? It already has one for if the HDF5_ENABLE_DEBUG_APIS option is enabled, so re-using that makes sense instead of adding more maintenance burden with two options the target is pulling from.

Reply:
That's fine if you just want to extend the option instead of creating new options.

Reply:
Another thing we could do is create a "DEBUG_HDF5_APIS" var that is the list of defines to be used and just use that var in the generator line. Instead of the add_compile_definitions(xxxx) do a list (append ${DEBUG_HDF5_APIS} xxxx).
Then change the definitions generator to use ${DEBUG_HDF5_APIS}.